api/agent/lb_agent.go

```diff
@@ -179,6 +190,45 @@ func (a *lbAgent) Submit(callI Call) error {
 	return a.handleCallEnd(ctx, call, err, true)
 }
 
+// setRequestGetBody sets GetBody function on the given http.Request if it is missing.  GetBody allows
+// reading from the request body without mutating the state of the request.
+func (a *lbAgent) setRequestBody(ctx context.Context, call *call) error {
+
+	r := call.req
+	if r.Body == nil || r.GetBody != nil {
+		return nil
+	}
+
+	// WARNING: we need to handle IO in a separate go-routine below
+	// to be able to detect a ctx timeout. When we timeout, we
+	// let gin/http-server to unblock the go-routine below.
+	errApp := make(chan error, 1)
+	go func() {
+
+		buffer, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			errApp <- err
+			return
+		}
+
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(buffer))
+
+		// GetBody does not mutate the state of the request body
+		r.GetBody = func() (io.ReadCloser, error) {
+			return ioutil.NopCloser(bytes.NewBuffer(buffer)), nil
```

Comment:
Also mentioned this in another review, but this seems to violate the contract of NewBuffer.  According to the documentation the Buffer object created on line 214 is the owner of `buffer`.  I understand that we're doing this to avoid a copy, but it makes me nervous.